File: leetcode/algorithms/541_reverse_string_2/main.java
class ReverseStringII {
    public String reverseStr(String s, int k) {
        if (s.length() < k) {
            return new StringBuffer(s).reverse().toString();
        }

        String result = "";
        for (int i = 0; i < s.length(); i += 2 * k) {
            String stringK = "";
            String string2K = "";
            if (i + k <= s.length()) {
                stringK = s.substring(i, i + k);
                if (i + 2 * k <= s.length()) {
                    string2K = s.substring(i + k, i + 2 * k);
                } else {
                    string2K = s.substring(i + k);
                }
            } else {
                stringK = s.substring(i);
            }
            result += new StringBuffer(stringK).reverse().toString() + string2K;
        }

        return result;
    }


    // Best Solution
    public String bestSolution1(String s, int k) {
        char[] arr = s.toCharArray();
        int n = arr.length;
        int i = 0;
        while (i < n) {
            int j = Math.min(i + k - 1, n - 1);
            swap(arr, i, j);
            i += 2 * k;
        }
        return String.valueOf(arr);
    }

    private void swap(char[] arr, int l, int r) {
        while (l < r) {
            char temp = arr[l];
            arr[l++] = arr[r];
            arr[r--] = temp;
        }
    }


    public String bestSolution2(String s, int k) {
        char[] ca = s.toCharArray();
        for (int left = 0; left < ca.length; left += 2 * k) {
            for (int i = left, j = Math.min(left + k - 1, ca.length - 1); i < j; i++, j--) {
                char tmp = ca[i];
                ca[i] = ca[j];
                ca[j] = tmp;
            }
        }
        return new String(ca);
    }
}

File: leetcode/algorithms/541_reverse_string_2/main.c
#include <string.h>

void swap(char* s, int l, int r) {
    while (l < r) {
        char temp = s[l];
        s[l++] = s[r];
        s[r--] = temp;
    }
}

char* reverseStr(char* s, int k) {
    int n = strlen(s);
    int i = 0;

    while (i < n) {
        int j = i + k - 1 < n - 1 ? i + k - 1 : n - 1;
        swap(s, i, j);
        i += 2 * k;
    }

    return s;
}


// Solution
void reverse_fn(char* s, int ind1, int ind2) {
    for (int i = 0; i < ((ind2 - ind1 + 1) / 2); i++) {
        char temp = s[ind1 + i];
        s[ind1 + i] = s[ind2 - i];
        s[ind2 - i] = temp;
    }
}

char* solution(char* s, int k) {
    int n, l, i;

    l = strlen(s);
    n = l / k;

    for (i = 1; i <= n; i++) {
        if (i % 2 == 1) {
            reverse_fn(s, (i - 1) * k, i * k - 1);
        }
    }

    if (l % k != 0) {
        if (n % 2 == 0) {
            reverse_fn(s, n * k, ((n * k) + (l % k) - 1));
        }
    }

    return s;
}
